lib/SILOptimizer/Utils/ConstExpr.cpp

```diff
@@ -0,0 +1,1854 @@
+
+  // Otherwise, we don't know how to handle this builtin.
+  return unknownResult();
+}
+
+// Handle calls to opaque callees, either by handling them and returning None or
+// by returning with a Unknown indicating a failure.
+llvm::Optional<SymbolicValue>
+ConstExprFunctionState::computeOpaqueCallResult(ApplyInst *apply,
+                                                SILFunction *callee) {
+  // This is a termination point like fatalError.
+  if (callee->hasSemanticsAttr("arc.programtermination_point")) {
+    // TODO: Actually get the message out of fatalError.  We can literally get
+    // its string and file/line/col info and propagate it up!
+    return evaluator.getUnknown((SILInstruction *)apply, UnknownReason::Trap);
+  }
+
+  if (classifyFunction(callee) == WellKnownFunction::AssertionFailure) {
+    // TODO: Actually get the message out of fatalError.  We can literally get
+    // its string and file/line/col info and propagate it up!
+    return evaluator.getUnknown((SILInstruction *)apply, UnknownReason::Trap);
+  }
+
+  LLVM_DEBUG(llvm::dbgs() << "ConstExpr Opaque Callee: " << *callee << "\n");
+  return evaluator.getUnknown((SILInstruction *)apply, UnknownReason::Default);
+}
+
+/// If the specified type is a Swift.Array of some element type, then return the
+/// element type.  Otherwise, return a null Type.
+static Type getArrayElementType(Type ty) {
+  if (auto bgst = ty->getAs<BoundGenericStructType>())
+    if (bgst->getDecl() == bgst->getASTContext().getArrayDecl())
+      return bgst->getGenericArgs()[0];
+  return Type();
+}
+
+// TODO: Refactor this to someplace common, this is defined in Devirtualize.cpp.
+SubstitutionMap getWitnessMethodSubstitutions(SILModule &Module, ApplySite AI,
+                                              SILFunction *F,
+                                              ProtocolConformanceRef CRef);
+
+/// Given a call to a function, determine whether it is a call to a constexpr
+/// function.  If so, collect its arguments as constants, fold it and return
+/// None.  If not, mark the results as Unknown, and return an Unknown with
+/// information about the error.
+llvm::Optional<SymbolicValue>
+ConstExprFunctionState::computeCallResult(ApplyInst *apply) {
+  auto conventions = apply->getSubstCalleeConv();
+
+  // Determine the callee.
+  auto calleeFn = getConstantValue(apply->getOperand(0));
+  if (calleeFn.getKind() != SymbolicValue::Function)
+    return evaluator.getUnknown((SILInstruction *)apply,
+                                UnknownReason::Default);
+
+  SILFunction *callee = calleeFn.getFunctionValue();
+
+  // If we reached an external function that hasn't been deserialized yet, make
+  // sure to pull it in so we can see its body.  If that fails, then we can't
+  // analyze the function.
+  if (callee->isExternalDeclaration()) {
+    callee->getModule().loadFunction(callee);
+    if (callee->isExternalDeclaration())
+      return computeOpaqueCallResult(apply, callee);
+  }
+
```

Comment:
@marcrasi 

Here is the compiler crash I found when using the example that uses string initializers. The interpreter runs fine but it results in a SIL verification failure in the later phases. The exact assertion failure is shown below, but at a high-level, this happens because the interpreter is pulling in the SIL of the stdlib function: `String.init(_builtinStringLiteral:utf8CodeUnitCount:isASCII:)` in the lines 783 to 790 (see the above code). The initializer calls `_StringGuts.init<A>(_large:)` which is left as an external function (without a body). The SILVerifier doesn't like this for some reason and fails. 

A possible solution
-------------------

I found that one way to fix this is to move the code in lines: 783 to 790 that loads an external function after the switch-case that handles well-known functions (i.e, after line 871 in the code that follows). This means that the SIL of an external function will not be loaded for a `WellKnownFunction` that the interpreter will handle specially. Does it makes sense? It seems this will eliminate some unnecessary work of loading the SIL of a well known function, and will also fix this assertion failure. 

Test case: 
----------

      struct Pair {
           var arr: [Int]
           var s: String
      }

      func bar() -> Pair {
          return Pair(s: "", x: 0)
      }

      func foo() {
        #assert(bar().x == 0) // assertion should succeed
      } 

Assertion failure
-----------------

`SIL verification failed: external declarations of SILFunctions with shared visibility is not allowed: SingleFunction || !hasSharedVisibility(RefF->getLinkage()) || RefF->hasForeignBody()`

`Verifying instruction:`
->   `// function_ref specialized _StringGuts.init<A>(_large:)
  %88 = function_ref @$ss11_StringGutsV6_largeABs010_UnmanagedA0VyxG_tcs17FixedWidthIntegerRzSURzlufCs5UInt8V_Tgq5 : $@convention(method) (_UnmanagedString<UInt8>, @thin _StringGuts.Type) -> @owned _StringGuts // user: %89`
    ` %89 = apply %88(%87, %52) : $@convention(method) (_UnmanagedString<UInt8>, @thin _StringGuts.Type) -> @owned _StringGuts // user: %90`

`In function:
// String.init(_builtinStringLiteral:utf8CodeUnitCount:isASCII:)`

Steps to reproduce
--------------------
`swiftc -frontend -emit-sil  -verify -enable-experimental-static-assert <string_test_case.swift>`